sklearn/preprocessing/label.py

```diff
@@ -33,6 +34,76 @@
 ]
 
 
+def _nanunique(ar):
+    uniques = np.unique(ar)
+    # np.nan is always sorted last
+    if is_scalar_nan(uniques[-1]):
+        nan_idx = np.searchsorted(uniques, np.nan)
+        return uniques[:nan_idx+1]
+    else:
+        return uniques
+
+
+def _nanencode_numpy(values, uniques=None, encode=False,
+                     missing_value=np.nan):
+    check_values = True
+    if uniques is None:
+        uniques = _nanunique(values)
+        u_mask = _get_mask(uniques, missing_value)
+        uniques = uniques[~u_mask]
+        check_values = False
+
+    if encode:
+        if check_values:
+            unique_v = _nanunique(values)
+            if is_scalar_nan(uniques[-1]) and is_scalar_nan(unique_v[-1]):
+                unique_v = unique_v[:-1]
+            unseen = np.setdiff1d(unique_v, uniques)
+            unseen = unseen[~_get_mask(unseen, missing_value)]
+            if unseen:
+                raise ValueError("y contains previously unseen labels: %s"
+                                 % str(unseen))
+
+        encoded = np.searchsorted(uniques, values)
+        e_mask = _get_mask(values, missing_value)
+        return uniques, encoded, e_mask
+    else:
+        return uniques
+
+
+def _nanencode_python(values, uniques=None, encode=False,
+                      missing_value=None):
+    if uniques is None:
+        uniques = set(values)
+        uniques.discard(missing_value)
+        # Put None at the end of the sort, similar to numpy.unique
+        uniques = sorted(uniques, key=lambda x: (x is None, x))
+        uniques = np.array(uniques, dtype=values.dtype)
+    if encode:
+        # Use index -1 so that the number encoding will not cause failures
+        # if used by the consumer for indexing. It will still fail when used
+        # for indexing an empty array but so is any other index.
+        nan_index = -1
+        table = {val: i for i, val in enumerate(uniques)}
+        table[missing_value] = nan_index
+        try:
+            encoded = np.array([table[v] for v in values])
```

Comment:
If `missing_values=nan` this won't work fully because `float('nan')` is neither identical nor equal to another `float('nan')`. Messy!

A relatively efficient way to handle this case might be to use `table.setdefault(v, -1)` in place of `table[v]`. Then after the encoding do something like:
```
unseen = [k for k, v in table.items() if v == -1]
if missing_values is None:
    unseen = [k for k in unseen if k is not None]
elif is_scalar_nan(missing_values):
    unseen = ... similar ...
else:
    unseen = ... similar ...
if unseen:
    raise ValueError("y contains previously unseen labels: %s" % unseen)
```

Reply:
It should work since dictionary key comparisons are done by hashes and not by values. That being said, ```np.nan```, ```float('nan')```, and ```np.float(`nan`)``` have different hashes.

Though I think using the np.nan is a more robust solution so I'll try to change it anyway.

Reply:
Dict lookups are done by hash, identity and equality. By their nature, a hash is not sufficient to match.